src/core/dxf/qgsdxfexport.cpp

```diff
@@ -1682,39 +1668,31 @@ void QgsDxfExport::addFeature( QgsSymbolRenderContext &ctx, const QgsCoordinateT
         }
 
         const QgsCurvePolygon *polygon = dynamic_cast<const QgsCurvePolygon *>( sourceGeom );
-        Q_ASSERT( polygon );
-
-        writePolyline( *polygon->exteriorRing(), layer, lineStyleName, penColor, width );
-        for ( int i = 0; i < polygon->numInteriorRings(); i++ )
-          writePolyline( *polygon->interiorRing( i ), layer, lineStyleName, penColor, width );
-
-        break;
-      }
-
-      case QgsWkbTypes::MultiSurface:
-      case QgsWkbTypes::MultiPolygon:
-      {
-        if ( !qgsDoubleNear( offset, 0.0 ) )
+        if ( polygon )
         {
-          QgsGeos geos( sourceGeom );
-          tempGeom.reset( geos.buffer( offset, 0, Qgis::EndCapStyle::Flat, Qgis::JoinStyle::Miter, 2.0 ) );  //#spellok
-          if ( tempGeom )
-            sourceGeom = tempGeom.get();
-          else
-            sourceGeom = geom.constGet();
+          writePolyline( *polygon->exteriorRing(), layer, lineStyleName, penColor, width );
+          for ( int i = 0; i < polygon->numInteriorRings(); i++ )
+            writePolyline( *polygon->interiorRing( i ), layer, lineStyleName, penColor, width );
         }
-
-        const QgsGeometryCollection *gc = dynamic_cast<const QgsGeometryCollection *>( sourceGeom );
-        Q_ASSERT( gc );
-
-        for ( int i = 0; i < gc->numGeometries(); i++ )
+        else
         {
-          const QgsCurvePolygon *polygon = dynamic_cast<const QgsCurvePolygon *>( gc->geometryN( i ) );
-          Q_ASSERT( polygon );
+          const QgsGeometryCollection *gc = dynamic_cast<const QgsGeometryCollection *>( sourceGeom );
+          if ( gc )
+          {
+            for ( int i = 0; i < gc->numGeometries(); i++ )
+            {
+              const QgsCurvePolygon *polygon = dynamic_cast<const QgsCurvePolygon *>( gc->geometryN( i ) );
+              Q_ASSERT( polygon );
 
-          writePolyline( *polygon->exteriorRing(), layer, lineStyleName, penColor, width );
-          for ( int j = 0; j < polygon->numInteriorRings(); j++ )
-            writePolyline( *polygon->interiorRing( j ), layer, lineStyleName, penColor, width );
+              writePolyline( *polygon->exteriorRing(), layer, lineStyleName, penColor, width );
+              for ( int j = 0; j < polygon->numInteriorRings(); j++ )
+                writePolyline( *polygon->interiorRing( j ), layer, lineStyleName, penColor, width );
+            }
+          }
+          else
+          {
+            Q_ASSERT( gc );
```

Comment:
That will always fail (so is equal to the previous check directly after the cast?

Reply:
I've moved it just after the cast